### PR TITLE
[Feature] Showing an alert when new conversation can't be added

### PR DESCRIPTION
**Pull request by KaterinaWire**
## What's new in this PR?
When we receive error code 412 in response to posting a new conversation, an alert will be shown with an error message.

Needs releases with:

- [ wireapp/wire-ios-sync-engine/pull/1443 ]

### CODE REVIEW

File: Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift [hunk 1]
```diff
@@ -34,13 +34,12 @@ extension UIAlertController {
 extension UIAlertController {
 
     @objc
-    func showErrorAlertForConnectionRequest() {
-        typealias ErrorString = L10n.Localizable.Error
-
-        let alertController = UIAlertController(title: ErrorString.Conversation.title,
-                                                message: ErrorString.Connection.missingLegalholdConsent,
+    public static func showErrorAlert(message: String) {
+        let alertController = UIAlertController(title: "error.conversation.title".localized,
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift:38]:
```suggestion
        let alertController = UIAlertController(title: L10n.Localizable.Error.Conversation.title,
```

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ConversationListContentController.swift [hunk 2]
```diff
@@ -96,6 +99,12 @@ final class ConversationListContentController: UICollectionViewController, Popov
         }
     }
 
+    @objc
+    func showErrorAlertForConversationRequest() {
+        typealias ErrorString = L10n.Localizable.Error
+        UIAlertController.showErrorAlert(message: ErrorString.Conversation.missingLegalholdConsent)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ConversationListContentController.swift:105]: The typealias isn't so necessary when you only use it once.

```suggestion
        UIAlertController.showErrorAlert(message: L10n.Localizable.Error.Conversation.missingLegalholdConsent)
```

File: Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift [hunk 3]
```diff
@@ -64,14 +64,20 @@ final class GroupDetailsViewController: UIViewController, ZMConversationObserver
...
 
     @available(*, unavailable)
     required init?(coder aDecoder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
 
+    @objc
+    func showErrorAlertForConnectionRequest() {
+        typealias ErrorString = L10n.Localizable.Error
+        UIAlertController.showErrorAlert(message: ErrorString.Connection.missingLegalholdConsent)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift:78]:
```suggestion
        UIAlertController.showErrorAlert(message: L10n.Localizable.Error.Connection.missingLegalholdConsent)
```

File: Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift [hunk 4]
```diff
@@ -675,6 +675,10 @@ final class ZClientViewController: UIViewController {
                 animated: Bool,
                 completion: Completion? = nil) {
         dismissAllModalControllers(callback: { [weak self] in
+            guard !conversation.isDeleted,
+                  let _ = conversation.managedObjectContext else {
+                return
+            }
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift:681]:
```suggestion
            guard
                !conversation.isDeleted,
                conversation.managedObjectContext != nil
            else {
                return
            }
```

File: Wire-iOS/Sources/UserInterface/StartUI/StartUI/SearchResultsViewController.swift [hunk 5]
```diff
@@ -245,6 +245,12 @@ final class SearchResultsViewController: UIViewController {
         searchResultsView.emptyResultContainer.isHidden = !isResultEmpty
     }
 
+    @objc
+    func showErrorAlertForConnectionRequest() {
+        typealias ErrorString = L10n.Localizable.Error
+        UIAlertController.showErrorAlert(message: ErrorString.Connection.missingLegalholdConsent)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/StartUI/StartUI/SearchResultsViewController.swift:251]:
```suggestion
        UIAlertController.showErrorAlert(message: L10n.Localizable.Error.Connection.missingLegalholdConsent)
```

File: Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift [hunk 6]
```diff
@@ -34,13 +34,12 @@ extension UIAlertController {
 extension UIAlertController {
 
     @objc
-    func showErrorAlertForConnectionRequest() {
-        typealias ErrorString = L10n.Localizable.Error
-
-        let alertController = UIAlertController(title: ErrorString.Conversation.title,
-                                                message: ErrorString.Connection.missingLegalholdConsent,
+    public static func showErrorAlert(message: String) {
```
Comment (David-Henner) [inline Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift:37]: Is it necessary to be made `public` ?

Comment (David-Henner) [reply Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift:37]: Is it an alert specifically related to the conversation errors? If so, should it be named differently?

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ConversationListContentController.swift [hunk 7]
```diff
@@ -85,6 +85,9 @@ final class ConversationListContentController: UICollectionViewController, Popov
         token = NotificationCenter.default.addObserver(forName: .activeMediaPlayerChanged, object: nil, queue: .main) { [weak self] _ in
             self?.activeMediaPlayerChanged()
         }
+
+        NotificationCenter.default.addObserver(self, selector: #selector(showErrorAlertForConversationRequest), name: ZMConversation.missingLegalHoldConsentNotificationName, object: nil)
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ConversationListContentController.swift:89]: should the observer be removed at some point?

File: Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift [hunk 8]
```diff
@@ -64,14 +64,19 @@ final class GroupDetailsViewController: UIViewController, ZMConversationObserver
                 }
             }
         }
-        NotificationCenter.default.addObserver(self, selector: #selector(UIAlertController.showErrorAlertForConnectionRequest), name: ZMConnectionNotification.missingLegalHoldConsent, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(showErrorAlertForConnectionRequest), name: ZMConnectionNotification.missingLegalHoldConsent, object: nil)
```
Comment (David-Henner) [inline Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift:67]: same as above

File: Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift (hunk 6 above)
Comment (KaterinaWire) [reply Wire-iOS/Sources/Helpers/UIAlertController+LocalizedError.swift:37]: To handle error messages for conversations and connections

File: Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift (hunk 8 above)
Comment (KaterinaWire) [reply Wire-iOS/Sources/UserInterface/GroupDetails/GroupDetailsViewController.swift:67]: Yes it happens in 
```
deinit {
        NotificationCenter.default.removeObserver(self)
    }
```

**Review by johnxnguyen — APPROVED** [review]

**Review by David-Henner — APPROVED** [review]

**Pull request merged**